src/plugins/core/header_grouping.ts

```diff
@@ -0,0 +1,523 @@
+          (i) =>
+            (i >= group.start && i <= group.end) ||
+            this.getters.isHeaderHiddenByUser(cmd.sheetId, cmd.dimension, i)
+        );
+        if (willHideAllHeaders) {
+          return CommandResult.NotEnoughElements;
+        }
+
+        break;
+    }
+    return CommandResult.Success;
+  }
+
+  handle(cmd: CoreCommand) {
+    switch (cmd.type) {
+      case "CREATE_SHEET":
+        this.history.update("groups", cmd.sheetId, { ROW: [], COL: [] });
+        break;
+      case "GROUP_HEADERS":
+        this.groupHeaders(cmd.sheetId, cmd.dimension, cmd.start, cmd.end);
+        break;
+      case "UNGROUP_HEADERS": {
+        this.unGroupHeaders(cmd.sheetId, cmd.dimension, cmd.start, cmd.end);
+        break;
+      }
+      case "DUPLICATE_SHEET": {
+        const groups = deepCopy(this.groups[cmd.sheetId]);
+        this.history.update("groups", cmd.sheetIdTo, groups);
+        break;
+      }
+      case "DELETE_SHEET": {
+        const groups = { ...this.groups };
+        delete groups[cmd.sheetId];
+        this.history.update("groups", groups);
+        break;
+      }
+      case "ADD_COLUMNS_ROWS":
+        const addIndex = getAddHeaderStartIndex(cmd.position, cmd.base);
+        this.moveGroupsOnHeaderInsertion(cmd.sheetId, cmd.dimension, addIndex, cmd.quantity);
+        break;
+      case "REMOVE_COLUMNS_ROWS":
+        this.moveGroupsOnHeaderDeletion(cmd.sheetId, cmd.dimension, cmd.elements);
+        break;
+      case "TOGGLE_HEADER_GROUP": {
+        const sheetId = cmd.sheetId;
+        const group = this.findGroupWithStartEnd(cmd.sheetId, cmd.dimension, cmd.start, cmd.end);
+        if (group) {
+          group.isFolded
+            ? this.unfoldHeaderGroup(sheetId, cmd.dimension, group)
+            : this.foldHeaderGroup(sheetId, cmd.dimension, group);
+        }
+        break;
+      }
+      case "EXPAND_ALL_HEADER_GROUPS":
+      case "COLLAPSE_ALL_HEADER_GROUPS": {
+        const groups = this.getters.getHeaderGroups(cmd.sheetId, cmd.dimension);
+        for (const group of groups) {
+          cmd.type === "EXPAND_ALL_HEADER_GROUPS"
+            ? this.unfoldHeaderGroup(cmd.sheetId, cmd.dimension, group)
+            : this.foldHeaderGroup(cmd.sheetId, cmd.dimension, group);
+        }
+        break;
+      }
+      case "COLLAPSE_HEADER_GROUPS_IN_ZONE":
+      case "EXPAND_HEADER_GROUPS_IN_ZONE": {
+        const action = cmd.type === "EXPAND_HEADER_GROUPS_IN_ZONE" ? "expand" : "collapse";
+        const layers = this.getGroupsLayers(cmd.sheetId, cmd.dimension);
+        if (action === "collapse") {
+          layers.reverse();
+        }
+        const groups = layers.flat();
+        const start = cmd.dimension === "ROW" ? cmd.zone.top : cmd.zone.left;
+        const end = cmd.dimension === "ROW" ? cmd.zone.bottom : cmd.zone.right;
+
+        const groupsToToggle = new Set<HeaderGroup>();
+        for (let header = start; header <= end; header++) {
```

Comment:
Yeah range is slighly more readable, but we've seen that it sucks performance-wise. Here it doesn't really matter, but I'm for avoiding using range, because if we see `range` everywhere we will also use it for methods whose performances matters. 

Just my 2 cents, I can change it if you really want.